### PR TITLE
Add legalhold command to allow setting legal hold

### PR DESCRIPTION
**Pull request by poornas**
on object prefixes.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -203,9 +203,10 @@ github.com/minio/lsync v1.0.1/go.mod h1:tCFzfo0dlvdGl70IT4IAK/5Wtgb0/BrTmo/jE8pA
 github.com/minio/minio v0.0.0-20200118222113-b849fd7a756d h1:/i9HncNqZaHhRLwaReYT73wpU1XFiKYIfhwCQOrVcRg=
 github.com/minio/minio v0.0.0-20200118222113-b849fd7a756d/go.mod h1:f8MCueGjpibyBY0iVthLZPcrz310Dgwe/Knx+Dx/2+A=
 github.com/minio/minio-go/v6 v6.0.44/go.mod h1:qD0lajrGW49lKZLtXKtCB4X/qkMf0a5tBvN2PaZg7Gg=
+github.com/minio/minio-go/v6 v6.0.45-0.20200117140906-66cf57d21ba4 h1:zjzoKgiRJm7k0J+IXz36u+ltbbU/iRTBlbm5/vZFt1w=
```
Comment (harshavardhana) [inline go.sum]: can you run `go mod tidy` ? @poornas

File: cmd/lock-main.go [hunk 2]
```diff
@@ -141,10 +141,6 @@ func mainLock(ctx *cli.Context) error {
 	console.SetColor("Mode", color.New(color.FgCyan, color.Bold))
 	console.SetColor("Validity", color.New(color.FgYellow))
 
-	// Parse encryption keys per command.
-	_, err := getEncKeys(ctx)
-	fatalIf(err, "Unable to parse encryption keys.")
-
```
Comment (nitisht) [inline cmd/lock-main.go]: now sure how is this related.

Comment (poornas) [reply cmd/lock-main.go]: This is not related - however including in this PR as this code cleanup is so minor that it doesnt warrant a separate PR. Removing this piece of code because encryption keys are needed only if a Stat or Get operation is attempted on an SSE-C encrypted object. `mc lock` is at bucket level and doesn't require encryption keys.

Comment (kannappanr) [reply cmd/lock-main.go]: There are no encryption keys involved with that command.

File: cmd/legalhold-main.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+	// Quit early if urlStr does not point to an S3 server
+	switch clnt.(type) {
+	case *fsClient:
+		fatal(errDummy().Trace(), "Legal hold not supported for file system.")
+	}
+
+	alias, _, _ := mustExpandAlias(urlStr)
+	if !isRecursive {
+		sse := getSSE(urlStr, encKeyDB[alias])
```
Comment (kannappanr) [inline cmd/legalhold-main.go]: Why do we need `sse` to set `legalhold` ?

File: cmd/legalhold-main.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// main for retention command.
+func mainLegalHold(ctx *cli.Context) error {
+	console.SetColor("LegalHoldSuccess", color.New(color.FgGreen, color.Bold))
+	console.SetColor("LegalHoldPartialFailure", color.New(color.FgRed, color.Bold))
+	console.SetColor("LegalHoldMessageFailure", color.New(color.FgYellow))
+	args := ctx.Args()
+	// Parse encryption keys per command.
+	encKeyDB, err := getEncKeys(ctx)
```
Comment (kannappanr) [inline cmd/legalhold-main.go]: We shouldn't require encryption keys

File: cmd/legalhold-main.go (hunk 3 above)
Comment (poornas) [reply cmd/legalhold-main.go]: removed - stat doesn't add much value, so letting list fail if recursive flag is not passed

File: cmd/legalhold-main.go (hunk 4 above)
Comment (poornas) [reply cmd/legalhold-main.go]: removed

File: README.md [hunk 5]
```diff
@@ -18,6 +18,7 @@ sql       run sql queries on objects
 stat      stat contents of objects
 lock      set and get object lock configuration
 retention set object retention for objects with a given prefix
+legalhold set object legal hold for objects with a given prefix
```
Comment (kannappanr) [inline README.md]: should we remove `with a given prefix` just like we have for `cp`

File: cmd/legalhold-main.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	  $ {{.HelpName}} myminio/mybucket/prefix/obj.csv ON
+ `,
+}
+
+// Structured message depending on the type of console.
+type legalHoldCmdMessage struct {
+	LegalHold minio.LegalHoldStatus `json:"legalhold"`
+	URLPath   string                `json:"urlpath"`
+	Status    string                `json:"status"`
+	Err       error                 `json:"error"`
```
Comment (kannappanr) [inline cmd/legalhold-main.go]: we should add `omit empty` for error

Comment (poornas) [reply cmd/legalhold-main.go]: done

File: cmd/client-s3.go [hunk 7]
```diff
@@ -2089,6 +2089,19 @@ func (c *s3Client) PutObjectRetention(mode *minio.RetentionMode, retainUntilDate
 	return nil
 }
 
+// Set object legal hold for a given object.
+func (c *s3Client) PutObjectLegalHold(lhold *minio.LegalHoldStatus) *probe.Error {
+	bucket, object := c.url2BucketAndObject()
```
Comment (kannappanr) [inline cmd/client-s3.go]: should we check for either bucketname or objectname being empty?

Comment (poornas) [reply cmd/client-s3.go]: it is already being done in minio-go, it is not necessary to duplicate this check here

**Comment by kannappanr** [comment]
```
$ mc legalhold play/test/issue ON
Cannot set object legal hold status `/test/issue`.Bucket is missing ObjectLockConfiguration
Errors found while setting legal hold status on objects with prefix `play/test/issue`. 
```
Need space between `.` and `Bucket`. Also, for a regular file, we do not have to print another line.

**Comment by kannappanr** [comment]
```
$ mc legalhold play/test/issue ON
Cannot set object legal hold status `/test/issue`. Bucket is missing ObjectLockConfiguration
Errors found while setting legal hold status on objects with prefix `play/test/issue`. 
```
The second line in the above message should be removed, if it is just a single object. Also, the text should start with `mc: <ERROR>` as is the case with the following example.
```
 $ mc rm play/test/issue1
mc: <ERROR> Failed to remove `play/test/issue1`. Target object is not found 

```

**Comment by kannappanr** [comment]
```
$ mc legalhold play/test1/issue on
Object legal hold successfully set for prefix `play/test1/issue`.

```
This should not say `prefix` in the message

**Comment by kannappanr** [comment]
```
$ mc legalhold play/test1/iss off 
Object legal hold successfully set for prefix `play/test1/iss`.

```
object `iss` doesn't exist, but an object `issue` exists, this seems to do `recursive`, always

**Comment by kannappanr** [comment]
```
 $ mc legalhold play/test1/iss1 off 
Object legal hold successfully set for prefix `play/test1/iss1`.

```
Object `iss1` doesn't exist and there is also no prefix

**Comment by kannappanr** [comment]
```
 $ mc legalhold play/test/iss1 offsasd 
mc: <ERROR> invalid legal hold status '2': invalid argument.
```
What does `2` mean here?

File: cmd/legalhold-main.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	args := ctx.Args()
+
+	var urlStr string
+	var lhold *minio.LegalHoldStatus
+	switch l := len(args); l {
+	case 2:
+		urlStr = args[0]
+		h := minio.LegalHoldStatus(strings.ToUpper(args[1]))
+		if !h.IsValid() {
+			fatalIf(probe.NewError(errors.New("invalid argument")), "invalid legal hold status '%v'", l)
```
Comment (kannappanr) [inline cmd/legalhold-main.go]:
```suggestion
			fatalIf(probe.NewError(errors.New("invalid argument")), "invalid legal hold status '%v'", h)
```

File: cmd/legalhold-main.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	// Quit early if urlStr does not point to an S3 server
+	switch clnt.(type) {
+	case *fsClient:
+		fatal(errDummy().Trace(), "Legal hold not supported for file system.")
+	}
+
+	alias, _, _ := mustExpandAlias(urlStr)
+	var cErr error
+	errorsFound := false
+	for content := range clnt.List(isRecursive, false, false, DirNone) {
```
Comment (kannappanr) [inline cmd/legalhold-main.go]: if `isRecursive` is false, we can simply call `PutObjectLegalHold` function here. Also, setLegalHold can be implemented as part of `client` interface, that way we do not have to check if the client type is fs or s3.

File: cmd/legalhold-main.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+		newClnt, perr := newClientFromAlias(alias, content.URL.String())
+		if perr != nil {
+			errorIf(content.Err.Trace(clnt.GetURL().String()), "Invalid URL")
+			continue
+		}
+		probeErr := newClnt.PutObjectLegalHold(lhold)
+		if probeErr != nil {
+			errorsFound = true
+			printMsg(legalHoldCmdMessage{
```
Comment (kannappanr) [inline cmd/legalhold-main.go]: should be `ErrorIf`

File: cmd/legalhold-main.go (hunk 9 above)
Comment (poornas) [reply cmd/legalhold-main.go]: fixed

File: cmd/legalhold-main.go (hunk 10 above)
Comment (poornas) [reply cmd/legalhold-main.go]: fixed

Comment (kannappanr) [reply cmd/legalhold-main.go]: still see `printMsg`
